hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java

```diff
@@ -0,0 +1,376 @@
+ * This can cause visibility issues due to a missing happens-before relation.
+ * <p>
+ * The collector fetches the data via {@link PartitionSpecificRunnable} (which
+ * run on the partition threads) and stores them in a {@link ConcurrentHashMap}.
+ * This guarantees the visibility for the {@code ClusterMergeTask}.
+ * <p>
+ * The collector can be implemented for partitioned and non-partitioned data
+ * structures via the base classes {@link AbstractPartitionedContainerCollector}
+ * and {@link AbstractNonPartitionedContainerCollector}.
+ * <p>
+ * <b>Note:</b> The collector will retain the collected containers for the merge runnable.
+ * So {@link ManagedService#reset()} is not allowed to clear or destroy the containers.
+ * To ensure this the collector removes the link between the data structure and a collected container via:
+ * <ul>
+ * <li>{@link Iterator#remove()} in {@link CollectPartitionedContainerRunnable#collectContainers()}
+ * for partitioned data structures</li>
+ * <li>{@link Map#remove(Object)} in {@link AbstractNonPartitionedContainerCollector#getContainer(String)}
+ * for non-partitioned data structures</li>
+ * </ul>
+ * The cleanup of the containers itself is done via {@link #destroy(Object)} after the merge is done
+ * or if the container is not collected. Containers from backup partitions are directly cleaned via
+ * {@link #destroyBackup(Object)}.
+ *
+ * @param <C> container of the data structure
+ */
+@SuppressWarnings("WeakerAccess")
+public abstract class AbstractContainerCollector<C> {
+
+    private final ConcurrentMap<Integer, Collection<C>> containersByPartitionId
+            = new ConcurrentHashMap<Integer, Collection<C>>();
+
+    private final OperationExecutor operationExecutor;
+    private final IPartitionService partitionService;
+    private final SplitBrainMergePolicyProvider mergePolicyProvider;
+
+    private CountDownLatch latch;
+
+    protected AbstractContainerCollector(NodeEngine nodeEngine) {
+        this.operationExecutor = ((OperationServiceImpl) nodeEngine.getOperationService()).getOperationExecutor();
+        this.partitionService = nodeEngine.getPartitionService();
+        this.mergePolicyProvider = nodeEngine.getSplitBrainMergePolicyProvider();
+    }
+
+    /**
+     * Collects the containers from the data structure in a thread-safe way.
+     */
+    public final void run() {
+        if (isPartitioned()) {
+            // collect partitioned containers
+            int partitionCount = partitionService.getPartitionCount();
+            latch = new CountDownLatch(partitionCount);
+
+            for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+                operationExecutor.execute(new CollectPartitionedContainerRunnable(partitionId));
+            }
+        } else {
+            // collect non-partitioned containers
+            Collection<String> containerNames = containerNames();
```

Comment:
Maybe I'm wrong but it seems there might be a race here. We first collect the containers and then collect their data on the partition thread. In the meantime, a new container might get created or an existing one might get destroyed (if the partition operation was already being executed while the node state was reset).
We should probably just run partition operations, regardless if it's a partitioned container or not, and collect both the names and data on the partition thread.
The downside being that once on the partition thread, you need to collect only the containers that belong to that partition.

Reply:
Hmm, you might be right, yeah. This will also get rid of the concurrent list, which is just needed since there might be multiple runnables per partition. I'll give it a try!

Reply:
That was really a good one, the code is even simpler now :)